shared/Microsoft.Extensions.Buffers.Sources/MemoryPoolSlab.cs

```diff
@@ -66,15 +76,36 @@ protected void Dispose(bool disposing)
 
                 _isActive = false;
 
+                _disposedValue = true;
+            }
+        }
+
+        internal void Return(MemoryPoolBlock block)
+        {
+#if BLOCK_LEASE_TRACKING
+            Debug.Assert(block.Slab == this, "Returned block was not this slab");
+            Debug.Assert(block.IsLeased, $"Block being returned to pool twice: {block.Leaser}{Environment.NewLine}");
+            block.IsLeased = false;
+#endif
+            GC.SuppressFinalize(block);
+
+            var remaining = Interlocked.Decrement(ref _ownedBlocks);
```

Comment:
I'm concerned about leaking pinned arrays. Are you sure we would always end up inpinning them?

Reply:
Not if the block was never returned